### PR TITLE
Allow Torch models to adjust to input dims

### PR DESCRIPTION
**Pull request by gheinrich**
Provide user-defined network definition with new
parameter "inputShape". The shape may be used
to programmatically configure the layers of the
network.

Standard networks have been updated to work
with grayscale or colour images indifferently.

New tests have been added to test the new feature.

### CODE REVIEW

File: digits/model/images/classification/views.py [hunk 1]
```diff
@@ -265,9 +265,6 @@ def image_classification_model_classify_one():
     db_task = job.train_task().dataset.train_db_task()
     height = db_task.image_dims[0]
     width = db_task.image_dims[1]
-    if job.train_task().crop_size:
```
Comment (gheinrich) [inline digits/model/images/classification/views.py]: I had to remove those lines to ensure consistency between methods: this seems to be the only place where images are being cropped before being fed to the DL framework (mean image, images in database and lists of images in classify_many are not cropped).

Comment (gheinrich) [inline digits/model/images/classification/views.py]: I have pushed a separate #395 for that

File: tools/torch/main.lua [hunk 2]
```diff
@@ -313,50 +379,20 @@ if lfs.mkdir(paths.concat(opt.save)) then
...
-    logmessage.display(0,'found ' .. valSize .. ' images in train db' .. opt.validation)
-end
-
 -- validate "crop length" input parameter
-if opt.crop then
-    if opt.croplen > imageSizeY then
-        logmessage.display(2,'invalid crop length! crop length ' .. opt.croplen .. ' is less than image width ' .. imageSizeY)
+if opt.crop and inputTensorShape then
+    if opt.croplen > inputTensorShape[3] then
+        logmessage.display(2,'invalid crop length! crop length ' .. opt.croplen .. ' is less than image width ' .. inputTensorShape[3])
```
Comment (lukeyeager) [inline tools/torch/main.lua]: No need to throw an error here. You can just make `croplen = min(croplen, width, height)`. That seems to be what Caffe does.

**Comment by lukeyeager** [comment]
When testing LeNet, I get an error for these image types - 32x32x3, 32x30x3, 28x28x1.
```
/home/lyeager/torch/install/bin/luajit: /home/lyeager/torch/install/share/lua/5.1/nn/Linear.lua:50: size mismatch at /home/lyeager/torch/extra/cutorch/lib/THC/THCTensorMathBlas.cu:90
stack traceback:
[C]: in function 'addmm'
/home/lyeager/torch/install/share/lua/5.1/nn/Linear.lua:50: in function 'updateOutput'
/home/lyeager/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
/home/lyeager/digits/tools/torch/main.lua:630: in function 'Validation'
/home/lyeager/digits/tools/torch/main.lua:787: in main chunk
[C]: in function 'dofile'
...ager/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
[C]: at 0x00406670
```

File: tools/torch/main.lua (hunk 2 above)
Comment (gheinrich) [inline tools/torch/main.lua]: OK thanks.

**Comment by gheinrich** [comment]
Standard models (LeNet, AlexNet and GoogLeNet) were updated to handle colour or grayscale images indifferently however they still require input images to have the intended image size:
LeNet: 28x28
AlexNet and GoogLeNet: 256x256

**Comment by gheinrich** [comment]
On @lukeyeager's suggestion I have updated standard models to drop a more explicit error if input images don't match the expected scale. For example:
https://github.com/NVIDIA/DIGITS/pull/393/files#diff-e694958b00d57e3fb85544725030c396R10

**Comment by lukeyeager** [comment]
That works. I was hoping for an error message to show up on the jobs page rather than having to look through the logs, but this is still an improvement over the previous error message.
```
2015-11-06 10:08:37 [INFO ] Loading mean tensor from /raid/jobs/dev/20151106-100734-4268/mean.jpg file
2015-11-06 10:08:37 [INFO ] Loading label definitions from /raid/jobs/dev/20151106-100734-4268/labels.txt file
2015-11-06 10:08:37 [INFO ] found 10 categories
2015-11-06 10:08:37 [INFO ] creating data readers
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/train_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/train_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/train_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/train_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] found 80 images in train db/raid/jobs/dev/20151106-100734-4268/train_db
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/val_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/val_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/val_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] opening LMDB database: /raid/jobs/dev/20151106-100734-4268/val_db
2015-11-06 10:08:37 [INFO ] Image channels are 1, Image width is 28 and Image height is 27
2015-11-06 10:08:37 [INFO ] found 20 images in train db/raid/jobs/dev/20151106-100734-4268/val_db
2015-11-06 10:08:37 [INFO ] Loading network definition from /raid/jobs/dev/20151106-100835-ee0f/model
/home/lyeager/torch/install/bin/luajit: /raid/jobs/dev/20151106-100835-ee0f/model.lua:10: Network expects 28x28 images
stack traceback:
[C]: in function 'assert'
/raid/jobs/dev/20151106-100835-ee0f/model.lua:10: in function 'network_func'
/home/lyeager/digits/tools/torch/main.lua:269: in main chunk
[C]: in function 'dofile'
...ager/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
[C]: at 0x00406670
```

**Comment by gheinrich** [comment]
I think the error is shown on the job page however it is true that you might have to scroll right to locate the error. It would be better if the error was printed where the UI currently shows `ERROR: error code 1` in large red letters. I am trying to implement an error catcher to do that.

**Comment by lukeyeager** [comment]
Right. We can fix that in a later pull request as it's a more general problem than just this PR.

**Comment by gheinrich** [comment]
Thanks. I'll propose another PR to improve error reporting.

**Pull request merged**